### PR TITLE
Order of the Lion

### LINKED ISSUE #2216: CONTENT: Monster Ancient Lion Knight mini boss and spawn script
I did and am sharing the Ancient Lion Knight and his spawn file every 15 minutes.

data>monsters>create folder "bounac">create ancient_lion_knight.lua insert code:

```
local mType = Game.createMonsterType("Ancient Lion Knight")
local monster = {}

monster.description = "a Ancient Lion Knight"
monster.experience = 8100
monster.outfit = {
	lookType = 1317,
	lookHead = 57,
	lookBody = 76,
	lookLegs = 57,
	lookFeet = 79,
	lookAddons = 1,
	lookMount = 0
}

monster.health = 9100
monster.maxHealth = 9100
monster.race = "blood"
monster.corpse = 39011
monster.speed = 340
monster.summonCost = 0
monster.maxSummons = 0

monster.changeTarget = {
	interval = 4000,
	chance = 60
}

monster.strategiesTarget = {
	nearest = 70,
	health = 10,
	damage = 10,
	random = 10,
}

monster.flags = {
	summonable = false,
	attackable = true,
	hostile = true,
	convinceable = false,
	pushable = false,
	rewardBoss = true,
	illusionable = false,
	canPushItems = true,
	canPushCreatures = true,
	staticAttackChance = 70,
	targetDistance = 1,
	runHealth = 0,
	healthHidden = false,
	isBlockable = false,
	canWalkOnEnergy = false,
	canWalkOnFire = true,
	canWalkOnPoison = true,
	pet = false
}

monster.light = {
	level = 0,
	color = 0
}


monster.loot = {
		{id = "platinum coin", chance = 100000, maxCount = 4},
		{id = "Ham", chance = 100000, maxCount = 3},
		{id = "Broken Helmet",  chance = 5000},
		{id = "Bug Meat",  chance = 5000},
		{id = "Life Preserver",  chance = 5000},
		{id = "Plate Armor",  chance = 5000},
		{id = "Combat Knife",  chance = 10000},
		{id = "Fishbone",  chance = 10000},
		{id = "Knife",  chance = 10000},
		{id = "Studded Shield",  chance = 10000},
		{id = "2229",  chance = 10000}, --skull
		{id = "Dark Armor",  chance = 7000},
		{id = "Dark Shield",  chance = 4000},
		{id = "Dirty Cape",  chance = 8900},
		{id = "Dirty Fur",  chance = 5000},
		{id = "Dwarven Shield",  chance = 9500},
		{id = "Twigs",  chance = 7000},
		{id = "Lion Axe",  chance = 200},
		{id = "Lion Hammer",  chance = 200},
		{id = "Lion Longbow",  chance = 200},
		{id = "Lion Longsword",  chance = 200},
		{id = "Lion Plate",  chance = 200},
		{id = "Lion Wand",  chance = 200},
		{id = "Lion Shield",  chance = 200},
		{id = "Lion Spangenhelm",  chance = 200},
		{id = "Lion Spellbook",  chance = 200}
}

monster.attacks = {
	{name ="melee", interval = 2000, chance = 100, minDamage = 900, maxDamage = -1450, effect = CONST_ME_DRAWBLOOD},
	{name ="combat", interval = 2000, chance = 35, type = COMBAT_EARTHDAMAGE, minDamage = -980, maxDamage = -1100, range = 7, radius = 7, shootEffect = CONST_ME_EXPLOSIONAREA, target = true},
	{name ="combat", interval = 2000, chance = 10, type = COMBAT_LIFEDRAIN, minDamage = -700, maxDamage = -920, range = 7, effect = CONST_ME_SMALLCLOUDS, target = true}
	
}

monster.defenses = {
	defense = 110,
	armor = 80,
	{name ="combat", interval = 2000, chance = 15, type = COMBAT_HEALING, minDamage = 250, maxDamage = 700, effect = CONST_ME_MAGIC_BLUE, target = false}
	
}

monster.elements = {
	{type = COMBAT_PHYSICALDAMAGE, percent = 95},
	{type = COMBAT_ENERGYDAMAGE, percent = 0},
	{type = COMBAT_EARTHDAMAGE, percent = 0},
	{type = COMBAT_FIREDAMAGE, percent = 90},
	{type = COMBAT_LIFEDRAIN, percent = 100},
	{type = COMBAT_MANADRAIN, percent = 0},
	{type = COMBAT_DROWNDAMAGE, percent = 0},
	{type = COMBAT_ICEDAMAGE, percent = 90},
	{type = COMBAT_HOLYDAMAGE , percent = -105},
	{type = COMBAT_DEATHDAMAGE , percent = -110}
}

monster.immunities = {
	{type = "paralyze", condition = true},
	{type = "outfit", condition = false},
	{type = "invisible", condition = true},
	{type = "bleed", condition = false}
}

mType.onThink = function(monster, interval)
end

mType.onAppear = function(monster, creature)
	if monster:getType():isRewardBoss() then
		monster:setReward(true)
	end
end

mType.onDisappear = function(monster, creature)
end

mType.onMove = function(monster, creature, fromPosition, toPosition)
end

mType.onSay = function(monster, creature, type, message)
end

mType:register(monster)
```

data>globalevents>create folder "bounac"> create file ancientlionknight-bounac-spawn.lua insert code:
```
local config = {
    monsterName = 'Ancient Lion Knight',
    bossPosition = Position(32440, 32471, 6),
    centerPosition = Position(32440, 32471, 6),
    rangeX = 10,
    rangeY = 10
}

local function checkBoss(centerPosition, rangeX, rangeY, bossName)
    local spectators, spec = Game.getSpectators(centerPosition, false, false, rangeX, rangeX, rangeY, rangeY)
    for i = 1, #spectators do
        spec = spectators[i]
        if spec:isMonster() then
            if spec:getName() == bossName then
                return true
            end
        end
    end
    return false
end

local ancientlionknight = GlobalEvent("ancientlionknight")
function ancientlionknight.onThink(interval, lastExecution)
    if checkBoss(config.centerPosition, config.rangeX, config.rangeY, config.monsterName) then
        return true
    end

    local boss =
    Game.createMonster(config.monsterName, config.bossPosition, true, true)
    boss:setReward(true)
    return true
end

ancientlionknight:interval(900000)
ancientlionknight:register()
```

### PR DESCRIPTION
**Pull request by hyresu**
This contains the base for the order of the lion boss fights
We need TESTING and FEEDBACK for this, as its just a base right now, it may need some more adjustments, so feel free to tell me the bugs and things to improve.

The fight is as rl, the usurpers fight lions, as well lions vs usurper all around bounac.

TODO:

- [x] Add boat to bounac
- [x] Add action ids for everything added (in the map)
- [x] Add spawns to bounac

Maybe I should move the "targetPreferPlayer" to a strategy on itself.

### CODE REVIEW

**Comment by dex-89** [comment]
@hyresu Look this: https://otland.net/threads/12-40-new-lion-sanctum-werelions-werelioness-with-spawn.271855/

**Comment by adrianomengue** [comment]
@hyresu  

I tested your pull request and I have some issues to report.

1. Drume's corpse id is wrong.
2. Kesar's speed attribute is now = 0, should change to 250 or something.
3. After killing drume, I have waited something about 15-30 minutes inside of the boss pit and I was kicked out.
4. I tried to logout inside the boss pit and I got it, I guess it couldn't be done.


Hope you can fix it. Thanks ;-)

**Review by adrianomengue — CHANGES_REQUESTED** [review]
I tested your pull request and I have some issues to report.

1. Drume's corpse id is wrong.
3. After killing drume, I have waited something about 15-30 minutes inside of the boss pit and I was kicked out.
4. I tried to logout inside the boss pit and I got it, I guess it couldn't be done.
5. Duplicate storage value 
storage
duplicated
6. You can calk through Lion Knight
    **Possible fix method:** isPet attribute on Lion Knight is true, change it to false maybe.
7. Comparing the respawn location with global version, they are too separeted from each other, I guess it was supposed to have the whole 3 groups fighting at same time in the screen


Hope you can fix it. Thanks ;-)

**Comment by adrianomengue** [comment]
> I tested your pull request and I have some issues to report.
> 
> 1. Drume's corpse id is wrong.
> 2. Kesar's speed attribute is now = 0, should change to 250 or something.
> 3. After killing drume, I have waited something about 15-30 minutes inside of the boss pit and I was kicked out.
> 4. I tried to logout inside the boss pit and I got it, I guess it couldn't be done.
> 5. Duplicate storage value
>    storage
>    duplicated
> 6. You can calk through Lion Knight
>    **Possible fix method:** isPet attribute on Lion Knight is true, change it to false maybe.
> 7. Comparing the respawn location with global version, they are too separeted from each other, I guess it was supposed to have the whole 3 groups fighting at same time in the screen
> 
> Hope you can fix it. Thanks ;-)


### Fixes:
**3. and 7.**  **_(data\scripts\quests\the_order_of_lion\drume.lua)_**
```lua
local config = {
	lionPosition = {
		Position(32444, 32512, 7),
		Position(32449, 32516, 7),
		Position(32444, 32520, 7),
	},
	usurperPosition = {
		Position(32450, 32520, 7),
		Position(32444, 32516, 7),
		Position(32448, 32512, 7)
	},
	firstPlayerPosition = Position(32457, 32508, 6),
    centerPosition = Position(32439, 32523, 7), -- Center Room  
	exitPosition = Position(32453, 32503, 7), -- Exit Position
	newPosition = Position(32453, 32510, 7),
	rangeX = 22,
	rangeY = 16,
	timeToKill = 5, -- time in minutes to remove the player	
}	

local function RoomIsOccupied(centerPosition, rangeX, rangeY)
	local spectators = Game.getSpectators(config.centerPosition, false, true, config.rangeX, config.rangeX, config.rangeY, config.rangeY)
	if #spectators ~= 0 then
		return true
	end

	return false
end


local function clearRoomDrume(centerPosition, rangeX, rangeY, resetGlobalStorage)
	local spectators,
	spectator = Game.getSpectators(centerPosition, false, false, rangeX, rangeX, rangeY, rangeY)
	for i = 1, #spectators do
		spectator = spectators[i]
		if spectator:isMonster() then
			spectator:remove()
		end
		if spectator:isPlayer() then
			spectator:teleportTo(config.exitPosition)
			spectator:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Your time is over.")
		end
	end
	if Game.getStorageValue(resetGlobalStorage) == 1 then
		Game.setStorageValue(resetGlobalStorage, -1)
	end
end

local drumeAction = Action()
function drumeAction.onUse(player, item, fromPosition, target, toPosition, isHotkey)
	if player:getPosition() ~= config.firstPlayerPosition then return false end
	if RoomIsOccupied(config.centerPosition, config.rangeX, config.rangeY) then
		player:sendCancelMessage("There's someone already in the skirmish.")
		player:getPosition():sendMagicEffect(CONST_ME_POFF)
		return true
	end
	local tempPos, tempTile, tempCreature
	local players = {}
	for x = config.firstPlayerPosition.x, config.firstPlayerPosition.x + 4 do
		tempPos = Position(x, config.firstPlayerPosition.y, config.firstPlayerPosition.z)
		tempTile = Tile(tempPos)
		if tempTile then
			tempCreature = tempTile:getTopCreature()
			if tempCreature and tempCreature:isPlayer() then
				table.insert(players, tempCreature)
			end
		end
	end
	if #players == 0 then return false end
	for _, pi in pairs(players) do
		if pi:getStorageValue(Storage.TheOrderOfTheLion.Drume.Timer) > os.time() then
			player:sendCancelMessage("Someone of your team has already fought in the skirmish in the last 20h.")
			player:getPosition():sendMagicEffect(CONST_ME_POFF)
			return true
		end
	end
	local spectators = Game.getSpectators(config.centerPosition, false, false, config.rangeX, config.rangeX, config.rangeY, config.rangeY)
	for _, creature in pairs(spectators) do
		if creature:isMonster() then
			creature:remove()
		end
	end
	local totalLion = 0
	local totalUsurper = 0
	local tempMonster
	for _, pos in pairs(config.lionPosition) do
		tempMonster = Game.createMonster("Lion Commander", pos)
		if not tempMonster then
			player:sendCancelMessage("There was an error, contact an admin.")
			player:getPosition():sendMagicEffect(CONST_ME_POFF)
			return true
		end
		totalLion = totalLion + 1
	end
	for _, pos in pairs(config.usurperPosition) do
		tempMonster = Game.createMonster("Usurper Commander", pos)
		if not tempMonster then
			player:sendCancelMessage("There was an error, contact an admin.")
			player:getPosition():sendMagicEffect(CONST_ME_POFF)
			return true
		end
		totalUsurper = totalUsurper + 1
	end
	for _, pi in pairs(players) do
		pi:setStorageValue(Storage.TheOrderOfTheLion.Drume.Timer, os.time() + (20 * 60 * 60))
		pi:teleportTo(config.newPosition)
		addEvent(clearRoomDrume, config.timeToKill * 60 * 1000, config.centerPosition, config.rangeX, config.rangeY, resetGlobalStorage)
		pi:sendTextMessage(MESSAGE_EVENT_ADVANCE, "You have " ..config.timeToKill.." minutes to defeat Drume.")
		
	end
	config.newPosition:sendMagicEffect(CONST_ME_TELEPORT)
	toPosition:sendMagicEffect(CONST_ME_POFF)
	Game.setStorageValue(Storage.TheOrderOfTheLion.Drume.TotalLionCommanders, totalLion)
	Game.setStorageValue(Storage.TheOrderOfTheLion.Drume.TotalUsurperCommanders, totalUsurper)
	return true
end
drumeAction:aid(59600)
drumeAction:register()
```
that's the new drume.lua fixed.

**4.** Just go to RME and cover the whole area with No Logout tile.
**6.** I changed the isPet attribute to false it seems to work well.


### Missing fixes:
>1 CorpseID should be the same as Lion Knight's corpse, but I don't know how to duplicate it.
>5 I'm not sure if I should change it, seems like they're intentionally the same.

Credits for who helped me: 
@dudantas

**Comment by adrianomengue** [comment]
Just another issue that I saw after:

Set all KESAR'S RESISTANCE to 100. He's immunity and set canWalkOnEnergy = false in Drume's config ;-)

File: src/combat.cpp [hunk 1]
```diff
@@ -335,7 +336,17 @@ ReturnValue Combat::canDoCombat(Creature* attacker, Creature* target)
...
 				}
 			}
+			if (attacker->getMonster() && (!attackerMaster || attackerMaster->getMonster())) {
+				if (attacker->getFaction() != FACTION_DEFAULT && !attacker->getMonster()->isEnemyFaction(targetPlayer->getFaction())) {
+					return RETURNVALUE_YOUMAYNOTATTACKTHISPLAYER;
+				}
+			}
 		} else if (target->getMonster()) {
+
+			if (attacker->getFaction() != FACTION_DEFAULT && attacker->getFaction() != FACTION_PLAYER && !attacker->getMonster()->isEnemyFaction(target->getFaction())) {
```
Comment (lgrossi) [inline src/combat.cpp]: attacker->getMonster() can be null

**Review by lgrossi — APPROVED** [review]

File: src/combat.cpp [hunk 2]
```diff
@@ -335,7 +336,17 @@ ReturnValue Combat::canDoCombat(Creature* attacker, Creature* target)
 					}
 				}
 			}
+			if (attacker->getMonster() && (!attackerMaster || attackerMaster->getMonster())) {
+				if (attacker->getFaction() != FACTION_DEFAULT && !attacker->getMonster()->isEnemyFaction(targetPlayer->getFaction())) {
+					return RETURNVALUE_YOUMAYNOTATTACKTHISPLAYER;
+				}
+			}
 		} else if (target->getMonster()) {
```
Comment (Costallat) [inline src/combat.cpp]:
```suggestion
		} else if (target && target->getMonster()) {
```

File: src/combat.cpp (hunk 1 above)
Comment (Costallat) [inline src/combat.cpp]:
```suggestion
			if (attacker->getFaction() != FACTION_DEFAULT && attacker->getFaction() != FACTION_PLAYER && attacker->getMonster() && !attacker->getMonster()->isEnemyFaction(target->getFaction())) {
```

File: src/monster.cpp [hunk 3]
```diff
@@ -476,9 +480,13 @@ bool Monster::isOpponent(const Creature* creature) const
 		if (creature != getMaster()) {
 			return true;
 		}
+	} else if (creature->getPlayer() && creature->getPlayer()->hasFlag(PlayerFlag_IgnoredByMonsters)) {
```
Comment (Costallat) [inline src/monster.cpp]:
```suggestion
	} else if (creature->getPlayer() &&  creature->getPlayer() && creature->getPlayer()->hasFlag(PlayerFlag_IgnoredByMonsters)) {
```

**Review by dudantas — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Pull request merged**